From 3804605cddd60c81d1656f987ca89e8064e5379a Mon Sep 17 00:00:00 2001
From: Murat Sezgin <msezgin@codeaurora.org>
Date: Thu, 28 Apr 2016 13:35:42 -0700
Subject: [PATCH] Pass serial number to Linux bridge port lookup

- This allows the same ECM lookup function to be used in interface
  hierarchy generation and when updating stats (so no difference in
  lookup even if the priority field changes)
- If the priority field changes, will update the H-Active entry and
  generate a debug log

Change-Id: I392bbc82fab30c345b14e7927b6b0fa119d19a57
Signed-off-by: Murat Sezgin <msezgin@codeaurora.org>
---
 include/linux/if_bridge.h | 6 ++++--
 net/bridge/br_if.c        | 7 ++++---
 2 files changed, 8 insertions(+), 5 deletions(-)

--- a/include/linux/if_bridge.h
+++ b/include/linux/if_bridge.h
@@ -72,7 +72,8 @@ int br_ioctl_call(struct net *net, struc
 		  struct ifreq *ifr, void __user *uarg);
 extern struct net_device *br_port_dev_get(struct net_device *dev,
 					  unsigned char *addr,
-					  struct sk_buff *skb);
+					  struct sk_buff *skb,
+					  unsigned int cookie);
 extern void br_refresh_fdb_entry(struct net_device *dev, const char *addr);
 extern void br_dev_update_stats(struct net_device *dev,
 				struct rtnl_link_stats64 *nlstats);
@@ -206,7 +207,8 @@ static inline clock_t br_get_ageing_time
 
 typedef struct net_bridge_port *br_port_dev_get_hook_t(struct net_device *dev,
 						       struct sk_buff *skb,
-						       unsigned char *addr);
+						       unsigned char *addr,
+						       unsigned int cookie);
 extern br_port_dev_get_hook_t __rcu *br_port_dev_get_hook;
 
 typedef void (br_notify_hook_t)(int group, int event, const void *ptr);
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -810,7 +810,8 @@ EXPORT_SYMBOL_GPL(br_port_flag_is_set);
  * associated port.
  */
 struct net_device *br_port_dev_get(struct net_device *dev, unsigned char *addr,
-				   struct sk_buff *skb)
+				   struct sk_buff *skb,
+				   unsigned int cookie)
 {
 	struct net_bridge_fdb_entry *fdbe;
 	struct net_bridge *br;
@@ -829,8 +830,8 @@ struct net_device *br_port_dev_get(struc
 		port_dev_get_hook = rcu_dereference(br_port_dev_get_hook);
 		if (port_dev_get_hook) {
 			struct net_bridge_port *pdst =
-				__br_get(port_dev_get_hook, NULL, dev,
-					skb, addr);
+				__br_get(port_dev_get_hook, NULL, dev, skb,
+					 addr, cookie);
 			if (pdst) {
 				dev_hold(pdst->dev);
 				netdev = pdst->dev;
